assets/js/modules/analytics/datastore/report.js

```diff
@@ -0,0 +1,151 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+import isPlainObject from 'lodash/isPlainObject';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from './constants';
+import { stringifyObject } from '../../../util';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+import {
+	isValidMetrics,
+	isValidDimensions,
+	isValidDateRange,
+	isValidOrders,
+} from '../../../util/report-validation';
+
+const fetchGetReportStore = createFetchStore( {
+	baseName: 'getReport',
+	controlCallback: ( { options } ) => {
+		return API.get( 'modules', 'analytics', 'report', options );
+	},
+	reducerCallback: ( state, report, { options } ) => {
+		return {
+			...state,
+			reports: {
+				...state.reports,
+				[ stringifyObject( options ) ]: report,
+			},
+		};
+	},
+	argsToParams: ( options ) => {
+		const {
+			metrics,
+			dimensions,
+			orderby,
+		} = options;
+
+		invariant( isPlainObject( options ), 'Options for Analytics report must be an object.' );
+		invariant( isValidDateRange( options ), 'Either date range or start/end dates must be provided for Analytics report.' );
+
+		if ( metrics ) {
+			invariant(
+				isValidMetrics( metrics ),
+				'Metrics for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "expression" and "alias" properties.',
+			);
+		}
+
+		if ( dimensions ) {
+			invariant(
+				isValidDimensions( dimensions ),
+				'Dimensions for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.',
+			);
+		}
+
+		if ( orderby ) {
+			invariant(
+				isValidOrders( orderby ),
+				'Orders for an Analytics report must be either an object or an array of objects where each object should have "fieldName" and "sortOrder" properties.',
+			);
+		}
+
+		return { options };
+	},
+} );
+
+const BASE_INITIAL_STATE = {
+	reports: {},
+};
+
+const baseResolvers = {
+	*getReport( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const existingReport = registry.select( STORE_NAME ).getReport( options );
+
+		// If there are already alerts loaded in state, consider it fulfilled
+		// and don't make an API request.
+		if ( existingReport ) {
+			return;
+		}
+
+		yield fetchGetReportStore.actions.fetchGetReport( options );
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Gets a Search Console report for the given options.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object}         state                     Data store's state.
+	 * @param {Object}         options                   Optional. Options for generating the report.
```

Comment:
This is required, because some of the options are required basically. We should probably not use `{}` (contrary to the ACs, I just noticed it's in there) as default, but instead fail in here if `! isPlainObject( options )`.

Reply:
Updated.